doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,256 @@
+# Universal, synchronous and in-thread loader hooks
+
+For background and motivation, see https://github.com/nodejs/node/issues/52219
+
+TL;DR: the top priority of this proposal is to allow sun-setting CJS loader monkey patching as non-breakingly as possible (which is why it needs to be synchronous and in-thread because that's how the CJS loader and `require()` works). Then it's API consistency with the existing `module.register()` off-thread hooks.
+
+Existing users of CJS loader monkey patching to look into:
+
+- pirates (used by Babel and nyc)
+- require-in-the-middle (used by many tracing agents)
+- yarn pnp
+- tsx
+- ts-node
+
+## API design
+
+A high-level overview of the API.
+
+```js
+const { addHooks, removeHooks } = require('module');
+
+function resolve(specifier, context, nextResolve) {
```

Comment:
What do you think about utilizing named parameters instead of positional?

```js
function resolve({ specifier, context, nextResolve })
```

Reply:
Since the `module.register` hooks use positional parameters it seems better to align with the existing API. I think with V8 there is a tiny overhead in the destructuring too since escape analysis with these can be difficult.